### ISSUE TITLE
DPOTrainer goes OOM after 300 steps

### ISSUE THREAD

**Issue by krypticmouse** [huggingface/trl#1087]
I'm trying to train the model on a ~12k-sized dataset with `DPOTrainer`. But the memory starts increasing from the first step and finally goes OOM around ~280 step. What am I doing wrong? Please help! I'm using a 80GB A100 GPU.

`DPOConfig` is the same as [here](https://github.com/huggingface/alignment-handbook/blob/61a11a5c7d66179ed0a930b0dd12e532fce701dd/src/alignment/configs.py#L245).

Please Help!

```python
    tokenizer = AutoTokenizer.from_pretrained(model_name)
    tokenizer.pad_token = tokenizer.eos_token

    model = AutoModelForCausalLM.from_pretrained(
        model_name,
        low_cpu_mem_usage=True,
        torch_dtype=torch.bfloat16,
        load_in_4bit=True,
        use_flash_attention_2=True,
        bnb_4bit_compute_dtype=torch.bfloat16,
        bnb_4bit_quant_type="nf4",
    )
    model.resize_token_embeddings(len(tokenizer))
    model.config.pad_token_id = tokenizer.pad_token_id
    model.config.use_cache = False
    
    ref_model = AutoModelForCausalLM.from_pretrained(
        model_name,
        low_cpu_mem_usage=True,
        torch_dtype=torch.bfloat16,
        load_in_4bit=True,
        use_flash_attention_2=True,
        bnb_4bit_compute_dtype=torch.bfloat16,
    ).eval()

    peft_config = LoraConfig(
        lora_alpha=128,
        lora_dropout=0.05,
        r=64,
        bias="none",
        task_type="CAUSAL_LM",
        target_modules=[
            "q_proj",
            "k_proj",
            "v_proj",
            "o_proj",
            "gate_proj",
            "up_proj",
            "down_proj",
            "lm_head",
        ],
    )
    model = get_peft_model(model, peft_config)

    training_args = DPOConfig(
        num_train_epochs=3,
        learning_rate=5e-07,
        per_device_train_batch_size=1,
        adam_epsilon=1e-08,
        lr_scheduler_type="linear",
        warmup_ratio=0.1,
        seed=42,
        logging_steps=100,
        save_steps=500,
        save_strategy="steps",
        output_dir="dpo-training-mistral",
        gradient_checkpointing=False,
        bf16=True,
        remove_unused_columns=False,
    )

    dpo_trainer = DPOTrainer(
        model,
        ref_model,
        args=training_args,
        beta=training_args.beta,
        train_dataset=dataset["train"],
        tokenizer=tokenizer,
        max_length=training_args.max_length,
        max_prompt_length=training_args.max_prompt_length,
        peft_config=peft_config,
    )

    dpo_trainer.train()
    dpo_trainer.save_model()
```

**Comment by younesbelkada** [comment]
hi @krypticmouse 
Thanks for the issue, can you try to enable gradient checkpointing? `gradient_checkpointing=True`
